File: LeetCode/0501-Find-Mode-in-BST/501.js
//Definition for a binary tree node.
function TreeNode(val, left, right) {
  this.val = val === undefined ? 0 : val;
  this.left = left === undefined ? null : left;
  this.right = right === undefined ? null : right;
}

/**
 * @param {TreeNode} root
 * @return {number[]}
 */

var findMode = function (root) {
    let count = 0, maxCount = 0;
    let pre = null, res = [];

    const traversal = (node) => {
        if (node === null) return;
        traversal(node.left);

        if (pre === null) {
            count = 1;
        } else if (pre.val === node.val) {
            count++;
        } else {
            count = 1;
        }
        pre = node;

        if (count === maxCount) res.push(node.val);
        if (count > maxCount) {
            maxCount = count;
            res = [];
            res.push(node.val);
        }
        traversal(node.right);
    };

    traversal(root);
    return res;

};

File: LeetCode/0501-Find-Mode-in-BST/501.cpp
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
	int val;
	TreeNode* left;
	TreeNode* right;
	TreeNode() : val(0), left(nullptr), right(nullptr) {}
	TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
	TreeNode(int x, TreeNode* left, TreeNode* right) : val(x), left(left), right(right) {}
};
/**
 * @brief
 * 若此题非 BST ，需要额外的 map 记录
 * BST 可用一个 pre 节点记录，与 cur 进行对比
 *
 */
class Solution {
private:
	int count = 0, maxCount = 0;
	vector<int> res;
	TreeNode* pre = nullptr;
	void traversal(TreeNode* node) {
		if (!node)
			return;
		traversal(node->left);

		if (!pre) {
			count = 1;
		} else if (pre->val == node->val) {
			++count;
		} else {
			count = 1;
		}
		pre = node;

		if (count == maxCount)
			res.push_back(node->val);

		if (count > maxCount) {
			res.clear();
			maxCount = count;
			res.push_back(node->val);
		}

		traversal(node->right);
	}

public:
	vector<int> findMode(TreeNode* root) {
		traversal(root);
		return res;
	}
};
int main() { return 0; }